From 1e70d23ffc7354fdb0aaea5ed6e1b25e639a1fb0 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Wed, 31 Mar 2021 16:00:42 +0300
Subject: [PATCH] [SW-37130] security: flash: modify spi_prot()

spi_prot() is modified as given in "WRSR flow"
in flash datasheet
@https://media.digikey.com/pdf/Data%20Sheets/Macronix/MX25U25635F.pdf

Now spi_prot() check added for 'WEL' bit to set before writing into
WRSR register. Also verified for WEL=0, BP[3:0], SPWD data
for successful WRSR update.

Also spi_prot() is moved to ppboot before pcie link init from
current spi_copy() function. Now as spi_prot() is moved before pci link
init, some whileloops are modified to use timeout to avoid infinite
loops.

Change-Id: I70a4b3161c99042b4491630927286e5b55f9ead9
Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
---

diff --git a/common/spi.c b/common/spi.c
index ea78d28..54fefea 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -15,6 +15,7 @@
 #include "common.h"
 #include <spi.h>
 #include "uart.h"
+#include <pll.h>
 
 /* Registers */
 #define CTRLR0				0x0
@@ -168,27 +169,33 @@
 	spi_release_bus();
 }
 
+
+/*
+ * 40ms as worst case WRSR cycle time given in mx25u25635f datasheet.
+ * Kept some extra(10ms) as timeout (total 50ms).
+ */
+#define SPI_WRITE_WAIT_TIMEOUT_US	50000
 static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
 {
 	unsigned char sr;
+	u64 start = timestamp_get_us();
 
 	spi_char_cmd_csel(CMD_WRITE_ENABLE, NULL);
-	spi_special_cmd_csel(cmd, buf, size);
-	do
+	do {
 		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	while (sr & READ_STATUS_WIP);
+		if (timeout_expired(start, SPI_WRITE_WAIT_TIMEOUT_US))
+			return;
+	} while (!(sr & READ_STATUS_WEL));
+
+	start = timestamp_get_us();
+	spi_special_cmd_csel(cmd, buf, size);
+	do {
+		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+		if (timeout_expired(start, SPI_WRITE_WAIT_TIMEOUT_US))
+			break;
+	} while (sr & READ_STATUS_WIP);
+
 	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
-}
-
-static unsigned char spi_prot(void)
-{
-	unsigned char sr;
-
-	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-	sr |= PROT_MASK;
-	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr));
-
-	return sr;
 }
 
 static void spi_read(void *dst, unsigned int offset, unsigned int size)
@@ -228,6 +235,30 @@
 	regmap_writel(0x00000001, spi_regs + SSIENR);
 }
 
+void spi_prot(void)
+{
+	unsigned char sr;
+
+	spi_init();
+
+	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+	sr |= PROT_MASK;
+	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr));
+	sr = 0; /* reset old value before read */
+	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
+
+	/* Avoided long string prints before pci link is enabled.
+	 * So added following numerical prints instead.
+	 * '6' means "Flash may be protected, depend on WP# signal"
+	 * '7' means "Flash is not protected"
+	 */
+	if ((sr & READ_STATUS_BP0_3) && (sr & READ_STATUS_SRWD)
+				&& !(sr & READ_STATUS_WEL))
+		preboot_puts("6\n");
+	else
+		preboot_puts("7\n");
+}
+
 void spi_copy_otp(void *dst, unsigned int offset, unsigned int size)
 {
 	spi_init();
@@ -238,19 +269,8 @@
 
 void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name)
 {
-	unsigned char sr;
-	static bool prot_flash = true;
-
 	spi_init();
 	spi_read(dst, offset, size);
 	preboot_puts_raw(name);
 	preboot_puts(" copying from flash is completed\n");
-	if (prot_flash) {
-		prot_flash = false;
-		sr = spi_prot();
-		if ((sr & READ_STATUS_BP0_3) && (sr & READ_STATUS_SRWD))
-			preboot_puts("Flash may be protected, depend on WP# signal\n");
-		else
-			preboot_puts("Flash is not protected\n");
-	}
 }
diff --git a/include/spi.h b/include/spi.h
index 3f5b267..55a31bc 100644
--- a/include/spi.h
+++ b/include/spi.h
@@ -12,6 +12,7 @@
 
 /* The Status Register (SR) in use bits definition */
 #define READ_STATUS_WIP			BIT(0)
+#define READ_STATUS_WEL			BIT(1)
 #define READ_STATUS_BP0			BIT(2)
 #define READ_STATUS_BP1			BIT(3)
 #define READ_STATUS_BP2			BIT(4)
@@ -29,5 +30,6 @@
 void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name);
 void spi_copy_otp(void *dst, unsigned int offset, unsigned int size);
 unsigned int get_cs_pad_sel(void);
+void spi_prot(void);
 
 #endif /* _HL_SPI_H_ */
diff --git a/ppboot/ppboot.c b/ppboot/ppboot.c
index 5b43b71..f56ee1c 100644
--- a/ppboot/ppboot.c
+++ b/ppboot/ppboot.c
@@ -251,6 +251,7 @@
 
 	preboot_sync_btl();
 	hl_uart_init();
+	spi_prot();
 	ppboot_pcie_init();
 	ppboot_pcie_drain_init();
 	ppboot_version();
